sycl/doc/SYCLCompilerAndRuntimeDesign.md

```diff
@@ -395,7 +395,35 @@ llvm-no-spir-kernel host.bc
 
 It returns 0 if no kernels are present and 1 otherwise.
 
-
+#### Device code splitting
+
+Putting all device code into a single SPIRV module does not work well in the
+following cases:
+1. There are thousands of kernels defined and only small part of them is used at
+run-time. Having them all in one SPIR-V module significantly increases JIT time.
+2. Device code can be specialized for different devices. For example, kernels
+which are supposed to be executed only on CPU can contain standard C library calls
+like "malloc" unavaliable on GPU. This will cause JIT compilation failure on GPU
+even if this particular kernel is never called on GPU device.
+
+To resolve these problems the compiler should be able to split a single module
+into smaller ones. The following features should be supported:
+* Grouping kernels on per-translation unit basis
+* Emitting a separate module for each kernel
+
+The current approach is:
+* Generate special meta-data with translation unit ID for each kernel in SYCL
+front-end
+* Link all LLVM modules using llvm-link
+* Run the module splitter pass on fully linked device code
```

Comment:
> Approach with using llvm-link to build the table of dependencies and importing them w/o doesn't sound flexible for extending kernels grouping scheme. For example we will still need some splitting mechanism if we decide leave one kernel per module.

If user cares about separating kernels into stand-alone modules, it can be done by following "one kernel per translation unit" code organization rule. No compiler changes are required for that.

> So, with using `llvm-link` options we still need figure out how to tell `llvm-link` which symbols we want to import for each module.
> Or run `llvm-link -only-needed` for all combinations of device modules which doesn't seem effective. And we will still need generate info about symbols which each module contains to feed it to `clang-offload-wrapper`.

We can link all module into one big module upfront and use `llvm-link -only-needed` to import all the necessary dependencies. Although this doesn't sound optimal neither it doesn't require new passes - just changes to the driver, which seems to be needed for original proposed too.

Reply:
> > Approach with using llvm-link to build the table of dependencies and importing them w/o doesn't sound flexible for extending kernels grouping scheme. For example we will still need some splitting mechanism if we decide leave one kernel per module.
> 
> If user cares about separating kernels into stand-alone modules, it can be done by following "one kernel per translation unit" code organization rule. No compiler changes are required for that.
> 
> > So, with using `llvm-link` options we still need figure out how to tell `llvm-link` which symbols we want to import for each module.
> > Or run `llvm-link -only-needed` for all combinations of device modules which doesn't seem effective. And we will still need generate info about symbols which each module contains to feed it to `clang-offload-wrapper`.
> 
> We can link all module into one big module upfront and use `llvm-link -only-needed` to import all the necessary dependencies. Although this doesn't sound optimal neither it doesn't require new passes - just changes to the driver, which seems to be needed for original proposed too.

It's unclear how to import all the necessary dependencies from fully linked module. Could you please clarify?
I'm thinking of reusing `llvm-extract` tool or re-using GVExtractor pass, these  solves problem of extraction of functions with dependencies from some single big module.

Reply:
> Did you consider using llvm-link to build the table of dependencies and importing them w/o "linking all LLVM modules using llvm-link"?

I think the potential a solution based on llvm-link building export/import tables and their analysis can be considered, but at this point it is too raw, and we don't even know will it work or not.
Doing full llvm-link is, in contrast, known to give desired result, and the design based on it is very simple. So I think we should start with it. Besides, llvm-link is very fast, so it is not quite clear what the potential optimizations will give.

> Yes, linking all device modules then split them back to separate modules sounds strange

This is not quite "splitting back". This is rather re-grouping based on callgraph knowledge, which produces very different modules compared to what was the input. Reliable building the callgraph based on just import/export tables w/o linking would be very tricky, I suppose.